omod/src/main/java/org/openmrs/module/owa/web/controller/OwaRestController.java

```diff
@@ -102,126 +106,146 @@
-					response.sendError(500, message);
+				checkInstall = this.installApp(uploadedFile, fileName, request, response, session);
+				if (checkInstall){
+					appManager.reloadApps();
+					appList = appManager.getApps();
+					return new ResponseEntity<List<App>>(appList, HttpStatus.OK);
 				}
+				return new ResponseEntity<List<App>>(HttpStatus.valueOf(500));
 			}
-			appManager.reloadApps();
-			appList = appManager.getApps();
+			return new ResponseEntity<List<App>>(appList, HttpStatus.BAD_REQUEST);
 		}
-		return appList;
+		return new ResponseEntity<List<App>>(HttpStatus.UNAUTHORIZED);
 	}
 
 	@RequestMapping(value = "/rest/owa/installapp", method = RequestMethod.POST)
 	@ResponseBody
-	public List<App> install(@RequestBody InstallAppRequestObject urlObject, HttpServletRequest request, HttpServletResponse response) throws IOException {
+	public ResponseEntity<List<App>> install(@RequestBody InstallAppRequestObject urlObject,
+	        HttpServletRequest request, HttpServletResponse response) throws IOException {
+		String message;
+		String fileName = null;
 		List<App> appList = new ArrayList<>();
 
-		String url = urlObject.getUrlValue();
-		String fileName = urlObject.getFileName();
-		if(fileName == null || fileName == ""){
-			throw new  NullPointerException(
-				"File name must be passed for installation"
-			);
-		}
-		URL downloadUrl = null;
-		if (ResourceUtils.isUrl(url)) {
-			downloadUrl = new URL(url);
-		}
 		if (Context.hasPrivilege("Manage OWA")) {
-			String message;
 			HttpSession session = request.getSession();
-			if (!url.isEmpty()) {
+			URL downloadUrl = null;
+			String installUrl = urlObject.getUrlValue();
+
+			if (!installUrl.contains(".zip")) {
+				message = messageSourceService.getMessage("Invalid URL");
+				log.warn("Invalid URL to OWA download");
+				session.setAttribute(WebConstants.OPENMRS_ERROR_ATTR, message);
+				response.sendError(400, message);
+				return new ResponseEntity<List<App>>(HttpStatus.BAD_REQUEST);
+			}
+
+			try {
+				downloadUrl = ResourceUtils.getURL(installUrl);
+				fileName = getPassedFileName(installUrl);
+				if (fileName == null || fileName.isEmpty()) {
+					message = messageSourceService.getMessage("Installation URL doesnot specify download file");
+					log.warn("No specified download file in Installation URL");
+					session.setAttribute(WebConstants.OPENMRS_ERROR_ATTR, message);
+					response.sendError(400, message);
+					return new ResponseEntity<List<App>>(HttpStatus.BAD_REQUEST);
+				}
 				InputStream inputStream = ModuleUtil.getURLStream(downloadUrl);
 				log.info("Url pathname: " + downloadUrl.getPath());
-				fileName += ".zip";
-				final String fileMatch=  fileName;
 				File file = ModuleUtil.insertModuleFile(inputStream, fileName);
-				try (ZipFile zip = new ZipFile(file)) {
-					if (zip.size() == 0) {
-						message = messageSourceService.getMessage("owa.blank_zip");
-						file.delete();
-						log.warn("Zip file is empty");
```

Comment:
Did you actually change all these? Or are they just formatting changes?

Reply:
since this action was being used both by install and upload. I separated it and created a method of it's own. This led to the multiple changes.